File: DFA.java
import java.util.*;

public class DFA {
    private Set<Integer> states;
    private Set<Character> alphabet;
    private Map<Integer, Map<Character, Integer>> transitions;
    private int startState;
    private Set<Integer> acceptStates;

    public DFA(Set<Integer> states, Set<Character> alphabet, Map<Integer, Map<Character, Integer>> transitions, int startState, Set<Integer> acceptStates) {
        this.states = states;
        this.alphabet = alphabet;
        this.transitions = transitions;
        this.startState = startState;
        this.acceptStates = acceptStates;
    }

    // 可视化DFA
    public void visualize() {
        System.out.println("DFA Visualization:");
        for (Map.Entry<Integer, Map<Character, Integer>> entry : transitions.entrySet()) {
            int state = entry.getKey();
            for (Map.Entry<Character, Integer> trans : entry.getValue().entrySet()) {
                char symbol = trans.getKey();
                int nextState = trans.getValue();
                System.out.println(state + " --" + symbol + "--> " + nextState);
            }
        }
    }

    // 扫描字符串
    public boolean scan(String input) {
        int currentState = startState;
        for (char symbol : input.toCharArray()) {
            if (transitions.containsKey(currentState) && transitions.get(currentState).containsKey(symbol)) {
                currentState = transitions.get(currentState).get(symbol);
            } else {
                return false;
            }
        }
        return acceptStates.contains(currentState);
    }
}

File: dfa.rs
use std::collections::{HashMap, HashSet};

pub struct DFA {
    states: HashSet<u32>,
    alphabet: HashSet<char>,
    transitions: HashMap<u32, HashMap<char, u32>>,
    start_state: u32,
    accept_states: HashSet<u32>,
}

impl DFA {
    pub fn new(
        states: HashSet<u32>,
        alphabet: HashSet<char>,
        transitions: HashMap<u32, HashMap<char, u32>>,
        start_state: u32,
        accept_states: HashSet<u32>,
    ) -> Self {
        DFA {
            states,
            alphabet,
            transitions,
            start_state,
            accept_states,
        }
    }

    // 可视化DFA
    pub fn visualize(&self) {
        println!("DFA Visualization:");
        for (state, trans) in &self.transitions {
            for (symbol, next_state) in trans {
                println!("{} --{}--> {}", state, symbol, next_state);
            }
        }
    }

    // 扫描字符串
    pub fn scan(&self, input: &str) -> bool {
        let mut current_state = self.start_state;
        for symbol in input.chars() {
            if let Some(trans) = self.transitions.get(&current_state) {
                if let Some(&next_state) = trans.get(&symbol) {
                    current_state = next_state;
                } else {
                    return false;
                }
            } else {
                return false;
            }
        }
        self.accept_states.contains(&current_state)
    }
}